### PR TITLE
Adding support for influx v3 in the influx scaler

### PR DESCRIPTION
**Pull request by danielwhatmuff**
* Implement required functionality to support influx v3
* Add tests
* Extend metadata parsing tests with additional data

Fixes https://github.com/kedacore/keda/issues/5445
Relates to https://github.com/kedacore/keda-docs/pull/1318

### CODE REVIEW

File: pkg/util/certificates.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ import (
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-const customCAPath = "/custom/ca"
+const customCAPath = "/tmp/custom/ca"
```
Comment (JorTurFer) [inline pkg/util/certificates.go:32]: Revert this change please

**Review by JorTurFer — COMMENTED** [review]
Looking good! Could you include an e2e tests for the new version?

File: pkg/util/certificates.go (hunk 1 above)
Comment (danielwhatmuff) [reply pkg/util/certificates.go:32]: will do, it was just for `make test` to work locally, didnt mean to push!

File: pkg/scalers/influxdb_scaler.go [hunk 2]
```diff
@@ -91,6 +131,8 @@ func parseInfluxDBMetadata(config *scalersconfig.ScalerConfig) (*influxDBMetadat
 	switch {
 	case ok && val != "":
 		organizationName = val
+	case config.TriggerMetadata["influxVersion"] == "3":
+		organizationName = val
```
[the remark this answers is not in the document]
Comment (danielwhatmuff) [reply pkg/scalers/influxdb_scaler.go:135]: organizations are not used in influx v3 (only databases) so its currently ignored if passed when using that version. Would it be better to use `organizationName = ""` here to make it more explicit (and keep it ignored) or throw an error if it's passed when using v3 with "organization not supported for Influx v3"?

**Comment by danielwhatmuff** [comment]
Something important to note - I do not have an influx v2 deployment available to perform any regression testing for that version.

**Comment by danielwhatmuff** [comment]
> Looking good! Could you include an e2e tests for the new version?

I'll take a look at these

File: pkg/util/certificates.go (hunk 1 above)
Comment (danielwhatmuff) [reply pkg/util/certificates.go]: done

**Comment by danielwhatmuff** [comment]
> > Looking good! Could you include an e2e tests for the new version?
> 
> I'll take a look at these

This isn't going to be possible as there's no public build for v3 available yet. It will be at some point later this year.

File: pkg/scalers/influxdb_scaler.go [hunk 3]
```diff
@@ -47,7 +61,29 @@ func NewInfluxDBScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
...
+		logger.Info("starting up influxdb v3 client")
+
+		clientv3, err := influxdb3.New(influxdb3.ClientConfig{
+			Host:     meta.serverURL,
+			Token:    meta.authToken,
+			Database: meta.database,
+		})
+
+		if err != nil {
+			panic(err)
```
Comment (zroubalik) [inline pkg/scalers/influxdb_scaler.go:74]: we should not panic here

File: pkg/scalers/influxdb_scaler.go [hunk 4]
```diff
@@ -47,7 +61,29 @@ func NewInfluxDBScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
 		return nil, fmt.Errorf("error parsing influxdb metadata: %w", err)
 	}
 
-	logger.Info("starting up influxdb client")
+	if meta.influxVersion == "3" {
+		logger.Info("starting up influxdb v3 client")
```
Comment (zroubalik) [inline pkg/scalers/influxdb_scaler.go:65]: this would be to verbose, let add `V(1)` here

File: pkg/scalers/influxdb_scaler.go [hunk 5]
```diff
@@ -47,7 +61,29 @@ func NewInfluxDBScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
...
+
+		return &influxDBScalerV3{
+			client:     *clientv3,
+			metricType: metricType,
+			metadata:   meta,
+			logger:     logger,
+		}, nil
+	}
+
+	logger.Info("starting up influxdb v2 client")
```
Comment (zroubalik) [inline pkg/scalers/influxdb_scaler.go:85]: this would be to verbose, let add `V(1)` here

**Review by zroubalik — COMMENTED** [review]
Shall we cover this also in e2e tests? @JorTurFer ?

File: pkg/scalers/influxdb_scaler.go (hunk 3 above)
Comment (danielwhatmuff) [reply pkg/scalers/influxdb_scaler.go]: updated, please let me know if it looks ok

File: pkg/scalers/influxdb_scaler.go (hunk 4 above)
Comment (danielwhatmuff) [reply pkg/scalers/influxdb_scaler.go]: updated

File: pkg/scalers/influxdb_scaler.go (hunk 5 above)
Comment (danielwhatmuff) [reply pkg/scalers/influxdb_scaler.go]: updated

File: pkg/scalers/influxdb_scaler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielwhatmuff) [reply pkg/scalers/influxdb_scaler.go:135]: happy to take any path here, just let me know

**Comment by JorTurFer** [comment]
> Shall we cover this also in e2e tests? @JorTurFer ?

yeah, we should

**Review by zroubalik — COMMENTED** [review]
Any update here?

**Comment by jmickey** [comment]
@zroubalik @JorTurFer Unfortunately, influxdata have yet to provide public images for influxdb v3. They've claimed this is eventually coming, both to us privately as customers and publicly here: https://www.influxdata.com/blog/the-plan-for-influxdb-3-0-open-source/. I don't think E2E testing is possible until this becomes reality, correct?

If it means anything, we've been running a version of Keda with these changes in production against our commercial Influxdb V3 version for a while now and it's been stable. Is there a path to get this merged without E2E tests against v3, possibly marking support as experimental, as long as the v2 tests continue to pass?

**Comment by JorTurFer** [comment]
> If it means anything, we've been running a version of Keda with these changes in production against our commercial Influxdb V3 installation for a while now and it's been stable. Is there a path to get this merged without E2E tests against v3, possibly marking support as experimental, as long as the v2 tests continue to pass?

We shouldn't merge any feature without the proper e2e tests. As the core is already public, can we just build our own image and use it temporally until they release their own one?
